docs/DafnyRef/UserGuide.md

```diff
@@ -190,9 +190,10 @@ Without a `Main` method, `dafny` will still produce executable output files, but
 they will need to be linked with some other code in the target language that
 provides a `main` entry point.
 
-TO BE WRITTEN
-
-TODO - non-static main in a class.
+The Main method may be declared in any module or class. If it is not static,
+then the containing class may not have any constructors.
+The method itself may not take arguments and may not have `requires` clauses
+in its specification.
```

Comment:
This can be overridden with the `{:main}` attribute, which allows the method to have any specification. (This is unsound in the same way that an `{:extern}` method is unsound. That is, the programmer takes responsibility for that the condition will really hold when the program starts.)

The code in `Compiler.IsMain` in the Compiler.cs allows `{:main}` to be placed on a method with any name. That seems to work for C#, but only if the method has no parameters. I tried it with Java and it crashes. So, apparently, the `{:main}` attribute is buggy. (I think the fix ought to be to look for `{:main}` only on a parameter-less method named `Main`, but allow it to have any specification.)

Reply:
This has been updated.